store/tikv/region_cache.go

```diff
@@ -74,76 +89,83 @@ func (c *RegionCache) NextPeer(id RegionVerID) {
 		return
 	}
 	if leader, err := region.NextPeer(); err != nil {
-		c.mu.Lock()
-		delete(c.regions, id)
-		c.mu.Unlock()
+		c.DropRegion(id)
 	} else {
 		c.UpdateLeader(id, leader.GetId())
 	}
 }
 
 // UpdateLeader update some region cache with newer leader info.
 func (c *RegionCache) UpdateLeader(regionID RegionVerID, leaderID uint64) {
-	old := c.GetRegionByVerID(regionID)
-	if old == nil {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	r, ok := c.regions[regionID]
+	if !ok {
 		log.Debugf("regionCache: cannot find region when updating leader %d,%d", regionID, leaderID)
 		return
 	}
-	var (
-		peer  *metapb.Peer
-		store *metapb.Store
-		err   error
-	)
 
-	curPeerIdx := -1
-	for idx, p := range old.meta.Peers {
+	var found bool
+	for i, p := range r.meta.Peers {
 		if p.GetId() == leaderID {
-			peer = p
-			// No need update leader.
-			if idx == old.curPeerIdx {
-				return
-			}
-			curPeerIdx = idx
+			r.curPeerIdx, r.peer = i, p
+			found = true
 			break
 		}
 	}
-	if peer != nil {
-		store, err = c.pdClient.GetStore(peer.GetStoreId())
+	if !found {
+		log.Debugf("regionCache: cannot find peer when updating leader %d,%d", regionID, leaderID)
+		c.dropRegionFromCache(r.VerID())
+		return
 	}
 
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	delete(c.regions, regionID)
-
-	if peer == nil || err != nil {
-		// Can't find the peer in cache, or error occurs when loading
-		// store from PD.
-		// Leave the region deleted, it will be filled later.
+	store, err := c.pdClient.GetStore(r.peer.GetStoreId())
+	if err != nil {
+		log.Warnf("regionCache: failed load store %d", r.peer.GetStoreId())
+		c.dropRegionFromCache(r.VerID())
 		return
 	}
 
-	c.regions[regionID] = &Region{
-		meta:       old.meta,
-		peer:       peer,
-		addr:       store.GetAddress(),
-		curPeerIdx: curPeerIdx,
-	}
+	r.addr = store.GetAddress()
 }
 
-// getRegionFromCache scan all region cache and find which region contains key.
 func (c *RegionCache) getRegionFromCache(key []byte) *Region {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-
-	for _, r := range c.regions {
-		if r.Contains(key) {
-			return r
-		}
+	var r *Region
+	c.sorted.DescendLessOrEqual(newRBSearchItem(key), func(item llrb.Item) bool {
+		r = item.(llrbItem).toRegion()
+		return false
+	})
+	if r == nil {
+		return nil
+	}
+	if r.Contains(key) {
+		return r.Clone()
 	}
 	return nil
 }
 
-// loadRegion get region from pd client, and pick the random peer as leader.
+func (c *RegionCache) insertRegionToCache(r *Region) {
+	if _, ok := c.regions[r.VerID()]; ok {
+		return
+	}
+	old := c.sorted.ReplaceOrInsert(newRBItem(r))
+	if old != nil {
+		delete(c.regions, old.(llrbItem).toRegion().VerID())
```

Comment:
@shenli It is more simple if we are sure that a Region contains our search key is in the cache. Besides, it won't matter if we use an old Region replace a new one, TiKV will reject the requests anyway.